packages/components/src/autocomplete/test/index.tsx

```diff
@@ -21,27 +22,45 @@ describe( 'AutocompleterUI', () => {
 
 			const resetSpy = jest.fn();
 
-			const autocompleter = {
+			const autocompleter: WPCompleter = {
 				name: 'fruit',
+				options: [
+					{ visual: '🍎', name: 'Apple', id: 1 },
+					{ visual: '🍊', name: 'Orange', id: 2 },
+					{ visual: '🍇', name: 'Grapes', id: 3 },
+				],
 				// The prefix that triggers this completer
 				triggerPrefix: '~',
+				getOptionLabel: ( option ) => (
+					<span>
+						<span className="icon">{ option.visual }</span>
+						{ option.name }
+					</span>
+				),
 				// Mock useItems function to return a autocomplete item.
-				useItems: () => {
-					return [
-						[
-							{
-								isDisabled: false,
-								key: 'Apple',
-								value: 'Apple',
-								label: (
-									<span>
-										<span className="icon">🍎</span>
-										{ 'Apple' }
-									</span>
-								),
-							},
-						],
-					];
+				useItems: ( filterValue ) => {
+					const options = autocompleter.options as Array< {
```

Comment:
The fact that we have an explicit `WPCompleter` annotation on line 25 _and_ an `as` assertion here makes me feel like something might be off. Like, something in this mock is wrong, or our types are wrong, or the DevEx could somehow be better.

For example, TS can already tell that `autocompleter.options` here is that fruit array we just assigned, but it's being thrown off because we explicitly annotated the broader `WPCompleter` type on line 25. It's not very elegant.

This isn't necessarily critical to address before merging, but it might be fun to figure out how you can make this work without an `as`. (I'd probably start by removing the `WPCompleter` annotation and work from there?)

Reply:
I agree with Lena — the fact that we need to "nudge" TS so often means poor DevX, and usually is the symptom that we could have typed our code better.

**I tried removing the `WPCompleter` type on line `25`** as Lena suggested, and worked my way through the errors:
- I added a `FuitOption` type to describe the objects in the `options` array, and used it to type the arguments of `getOptionLabel`
- in `useItems`:
  - I removed the type cast to `autocompleter.options`, since TS can now infer the type directly
  - I removed the ternary operators around `autocompleter.getOptionKeywords` and ` autocompleter.isOptionDisabled` — TS was complaining because effectively they are not defined on the `autocompleter` object.[ I also thought it was safe to remove them, since they don't seem to be used in the tests.](https://github.com/WordPress/gutenberg/pull/47751/files#r1099464037) Another option here would be to define them on the `autocompleter` object
  - I used `as const` to narrow the return type of the function, similarly to what I've suggested for `getDefaultUseItems`
  - as a consequence of this last change, I also had to mark the type of `WPCompleter['useItems']` as `readonly` (since `as const` adds `readonly` automatically)

<details>

<summary>Here's the full diff with those changes applied</summary>

```diff
diff --git a/packages/components/src/autocomplete/test/index.tsx b/packages/components/src/autocomplete/test/index.tsx
index 081e441d13..9322ad7a35 100644
--- a/packages/components/src/autocomplete/test/index.tsx
+++ b/packages/components/src/autocomplete/test/index.tsx
@@ -13,7 +13,12 @@ import { useRef } from '@wordpress/element';
  * Internal dependencies
  */
 import { getAutoCompleterUI } from '../autocompleter-ui';
-import type { WPCompleter } from '../types';
+
+type FruitOption = {
+	visual: string;
+	name: string;
+	id: number;
+};
 
 describe( 'AutocompleterUI', () => {
 	describe( 'click outside behavior', () => {
@@ -22,7 +27,7 @@ describe( 'AutocompleterUI', () => {
 
 			const resetSpy = jest.fn();
 
-			const autocompleter: WPCompleter = {
+			const autocompleter = {
 				name: 'fruit',
 				options: [
 					{ visual: '🍎', name: 'Apple', id: 1 },
@@ -31,36 +36,28 @@ describe( 'AutocompleterUI', () => {
 				],
 				// The prefix that triggers this completer
 				triggerPrefix: '~',
-				getOptionLabel: ( option ) => (
+				getOptionLabel: ( option: FruitOption ) => (
 					<span>
 						<span className="icon">{ option.visual }</span>
 						{ option.name }
 					</span>
 				),
 				// Mock useItems function to return a autocomplete item.
-				useItems: ( filterValue ) => {
-					const options = autocompleter.options as Array< {
-						visual: 'string';
-						name: 'string';
-						id: number;
-					} >;
+				useItems: ( filterValue: string ) => {
+					const options = autocompleter.options;
 					const keyedOptions = options.map(
 						( optionData, optionIndex ) => ( {
 							key: `${ autocompleter.name }-${ optionIndex }`,
 							value: optionData,
 							label: autocompleter.getOptionLabel( optionData ),
-							keywords: autocompleter.getOptionKeywords
-								? autocompleter.getOptionKeywords( optionData )
-								: [],
-							isDisabled: autocompleter.isOptionDisabled
-								? autocompleter.isOptionDisabled( optionData )
-								: false,
+							keywords: [],
+							isDisabled: false,
 						} )
 					);
 					const filteredOptions = keyedOptions.filter( ( option ) =>
 						option.value.name.includes( filterValue )
 					);
-					return [ filteredOptions ];
+					return [ filteredOptions ] as const;
 				},
 			};
 
diff --git a/packages/components/src/autocomplete/types.ts b/packages/components/src/autocomplete/types.ts
index a6a8d10790..a2441437c3 100644
--- a/packages/components/src/autocomplete/types.ts
+++ b/packages/components/src/autocomplete/types.ts
@@ -80,7 +80,7 @@ export type WPCompleter< TCompleterOption = any > = {
 	 * Autocomplete UI. These items have uniform shape and have been filtered by
 	 * `AutocompleterUIProps.filterValue`.
 	 */
-	useItems?: ( filterValue: string ) => [ Array< KeyedOption > ];
+	useItems?: ( filterValue: string ) => readonly [ Array< KeyedOption > ];
 	/**
 	 * Whether or not changes to the `filterValue` should be debounced.
 	 */
```

</details>

**I then tried the "opposite" approach** — keeping `WPComplete` on line `25` and trying to leverage it as much as possible.

- I noticed that `WPComplete` accepts a `TCompleterOption` generic — and so I decided to pass `FruitOption` as the value for the generic
- This still didn't make TS happy about `autocompleter.options` in the `useItems` function. That is because `WPCompleter['options']` can either be an `Array< TCompleterOption >`, but it can also be of type `GetOptions`. When writing a custom `useItems` function, TS cannot exclude the `GetOptions` possibility, and therefore it errors
- I still think that we can improve the types of options, by forwarding `TCompleterOption` to `GetOptions`
- Still, in `useItems` we should _technically_ handle the case in which options is of type `GetOptions< TCompleterOption >`.
- Since in this specific case we know that `options` is a play array, using the typecast `as FruitOption[]` is safe and avoids us from writing a lot of unnecessary code.

<details>

<summary>Here's a summary of the changes</summary>

```diff
diff --git a/packages/components/src/autocomplete/test/index.tsx b/packages/components/src/autocomplete/test/index.tsx
index 081e441d13..cba1cd91c0 100644
--- a/packages/components/src/autocomplete/test/index.tsx
+++ b/packages/components/src/autocomplete/test/index.tsx
@@ -15,6 +15,12 @@ import { useRef } from '@wordpress/element';
 import { getAutoCompleterUI } from '../autocompleter-ui';
 import type { WPCompleter } from '../types';
 
+type FruitOption = {
+	visual: string;
+	name: string;
+	id: number;
+};
+
 describe( 'AutocompleterUI', () => {
 	describe( 'click outside behavior', () => {
 		it( 'should call reset function when a click on another element occurs', async () => {
@@ -22,7 +28,7 @@ describe( 'AutocompleterUI', () => {
 
 			const resetSpy = jest.fn();
 
-			const autocompleter: WPCompleter = {
+			const autocompleter: WPCompleter< FruitOption > = {
 				name: 'fruit',
 				options: [
 					{ visual: '🍎', name: 'Apple', id: 1 },
@@ -39,11 +45,7 @@ describe( 'AutocompleterUI', () => {
 				),
 				// Mock useItems function to return a autocomplete item.
 				useItems: ( filterValue ) => {
-					const options = autocompleter.options as Array< {
-						visual: 'string';
-						name: 'string';
-						id: number;
-					} >;
+					const options = autocompleter.options as FruitOption[];
 					const keyedOptions = options.map(
 						( optionData, optionIndex ) => ( {
 							key: `${ autocompleter.name }-${ optionIndex }`,
diff --git a/packages/components/src/autocomplete/types.ts b/packages/components/src/autocomplete/types.ts
index a6a8d10790..25bc1411ed 100644
--- a/packages/components/src/autocomplete/types.ts
+++ b/packages/components/src/autocomplete/types.ts
@@ -43,7 +43,7 @@ export type WPCompleter< TCompleterOption = any > = {
 	 * Options may be of any type or shape. The completer declares how those
 	 * options are rendered and what their completions should be when selected.
 	 */
-	options: Array< TCompleterOption > | GetOptions;
+	options: Array< TCompleterOption > | GetOptions< TCompleterOption >;
 	/**
 	 * A function that returns the keywords for the specified option.
 	 */
```
</details>

----

Final thoughts:

- the cleanest approach seems to be the first one, in which I removed the `WPCompleter` type 
- working on the second approach was still interesting, as it makes me wonder if there's a better way to type `options`. Right now, the fact that it can be an array / or a function returning an array / or a function returning a promise for an array does not provide great DevX 🤔

Reply:
Wow, thanks for digging in here, both of you. I looked into a bit before the weekend but couldn't quite resolve the typing conflicts, and I see why now!

I've implemented @ciampo's first suggestion of removing the `WPCompleter` and using `const` in 1bd1cfc8d4

I then got to thinking about how to address they typing of `options` so custom `useItems` functions wouldn't run into this problem in 4bb27eecce. I _think_ that should allow TypeScript to parse the difference between an array of options and a function. Curious what you think!

Reply:
Your solution seems to go in the right direction!

Thanks to that, I tried again the `WPCompleter` approach, and I must say that it now feels like the cleaner option

<details>
<summary>We now only need to define the `WPComplete< FruitOption >`, and TSC is happy</summary>

```diff
diff --git a/packages/components/src/autocomplete/test/index.tsx b/packages/components/src/autocomplete/test/index.tsx
index e85e0ca91c..4be49f0f27 100644
--- a/packages/components/src/autocomplete/test/index.tsx
+++ b/packages/components/src/autocomplete/test/index.tsx
@@ -13,6 +13,7 @@ import { useRef } from '@wordpress/element';
  * Internal dependencies
  */
 import { getAutoCompleterUI } from '../autocompleter-ui';
+import type { WPCompleter } from '../types';
 
 type FruitOption = { visual: string; name: string; id: number };
 
@@ -23,7 +24,7 @@ describe( 'AutocompleterUI', () => {
 
 			const resetSpy = jest.fn();
 
-			const autocompleter = {
+			const autocompleter: WPCompleter< FruitOption > = {
 				name: 'fruit',
 				options: [
 					{ visual: '🍎', name: 'Apple', id: 1 },
@@ -32,14 +33,14 @@ describe( 'AutocompleterUI', () => {
 				],
 				// The prefix that triggers this completer
 				triggerPrefix: '~',
-				getOptionLabel: ( option: FruitOption ) => (
+				getOptionLabel: ( option ) => (
 					<span>
 						<span className="icon">{ option.visual }</span>
 						{ option.name }
 					</span>
 				),
 				// Mock useItems function to return a autocomplete item.
-				useItems: ( filterValue: string ) => {
+				useItems: ( filterValue ) => {
 					const options = autocompleter.options;
 					const keyedOptions = options.map(
 						( optionData, optionIndex ) => ( {
@@ -53,7 +54,7 @@ describe( 'AutocompleterUI', () => {
 					const filteredOptions = keyedOptions.filter( ( option ) =>
 						option.value.name.includes( filterValue )
 					);
-					return [ filteredOptions ] as const;
+					return [ filteredOptions ];
 				},
 			};

```
</details>

This approach also has the advantage that if in the future we want to reference `autocompleter.getOptionKeywords` or `autocompleter.isOptionDisabled`, TSC won't complain because it knows they are defined on a `WPCompleter` object.

Reply:
> I've implemented @ciampo's first suggestion of removing the `WPCompleter` and using `const` in [1bd1cfc](https://github.com/WordPress/gutenberg/commit/1bd1cfc8d45d76a1addcb6830e5ee1d9c63428f2)

Thank you!

> I then got to thinking about how to address they typing of `options` so custom `useItems` functions wouldn't run into this problem in [4bb27ee](https://github.com/WordPress/gutenberg/commit/4bb27eeccefe4fda4b4b9463a8de86496fa86e1f). I _think_ that should allow TypeScript to parse the difference between an array of options and a function. Curious what you think!

The reasoning behind your idea is correct, but your implementation won't work. `TCompleterOption` in fact represents the type of single option, and therefore we won't be able to assess if `options` is an array, a function or a promise just by looking at it

Reply:
Good point, I got ahead of myself there. Rolled back: 43adf01ae7